be/src/exec/spill/executor.h

```diff
@@ -14,28 +14,58 @@
 
 #pragma once
 
+#include <memory>
 #include <utility>
 
 #include "common/compiler_util.h"
+#include "exec/pipeline/pipeline_fwd.h"
+#include "exec/pipeline/query_context.h"
 #include "runtime/current_thread.h"
 #include "runtime/mem_tracker.h"
 #include "util/priority_thread_pool.hpp"
 
-namespace starrocks {
-namespace spill {
+namespace starrocks::spill {
 struct EmptyMemGuard {
-    void scoped_begin() const {}
+    bool scoped_begin() const { return true; }
     void scoped_end() const {}
 };
 
 struct MemTrackerGuard {
     MemTrackerGuard(MemTracker* scope_tracker_) : scope_tracker(scope_tracker_) {}
-    void scoped_begin() const { old_tracker = tls_thread_status.set_mem_tracker(scope_tracker); }
+    bool scoped_begin() const {
+        old_tracker = tls_thread_status.set_mem_tracker(scope_tracker);
+        return true;
+    }
     void scoped_end() const { tls_thread_status.set_mem_tracker(old_tracker); }
     MemTracker* scope_tracker;
     mutable MemTracker* old_tracker = nullptr;
 };
 
+struct QueryCtxMemTrackerGuard {
```

Comment:
suggest to implement EmptyMemGuard/MemTrackerGuard/QueryCtxMemTrackerGuard  as RAII idioms, 
use constructor and destructor instead of scoped_begin and scoped_end,  RAII is more robust.

Reply:
I've thought about this, mainly because of whether putting some scoped_begin logic in the memory guard's constructor would have unintended effects (because of the thread local variable settings involved). 

On the other hand, we may need to use MemGuard to end the function early, which we may not be able to do using RAII directly